orchestration/dagster_orchestration/hca_orchestration/config/preconfiguration_loader.py

```diff
@@ -0,0 +1,91 @@
+from dataclasses import dataclass
+import os
+from typing import Optional
+import warnings
+
+import yaml
+
+import hca_orchestration.config
+from hca_orchestration.support.typing import DagsterConfigDict
+
+
+@dataclass
+class PreconfigurationLoader:
+    name: str
+    subdirectory: str
+    optional_keys: set[str]
+    required_keys: set[str]
+
+    def __post_init__(self) -> None:
+        self.config_file_directory = os.path.join(
+            os.path.dirname(hca_orchestration.config.__file__),
+            self.subdirectory
+        )
+
+    def file_path(self, filename: str) -> str:
+        return os.path.join(self.config_file_directory, filename)
+
+    def load_config(self, filename: str) -> Optional[DagsterConfigDict]:
+        config_file_path = self.file_path(filename)
+        if os.path.exists(config_file_path):
+            with open(config_file_path, 'r') as config_file_io:
+                return DagsterConfigDict(yaml.safe_load(config_file_io))
+
+        return None
+
+    # raises an error for any missing config keys, records a warning for (and discards) any extra keys
+    def validated_config(self, config: DagsterConfigDict) -> DagsterConfigDict:
+        keys_present = set(config.keys())
+        permitted_keys = self.required_keys | self.optional_keys
+        missing_keys = self.required_keys - keys_present
+        extra_keys = keys_present - permitted_keys
+
+        if any(missing_keys):
+            missing_keys_str = ", ".join(key for key in missing_keys)
+            raise ValueError(
+                f"Missing expected preconfigured field(s) in configuration files for {self.name}. "
+                f"Config files do not define these required fields:\n{missing_keys_str}\n"
```

Comment:
👍 I like this error message, dagster does a good job of providing actionable msgs like this when misconfigurations occur and this is keeping with that pattern.